geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java

```diff
@@ -0,0 +1,236 @@
+  }
+
+  @After
+  public void teardown() {
+    springContext.stop();
+    jedis.flushAll();
+    sessionIds.clear();
+  }
+
+  @Test
+  public void sessionOperationsDoNotFail_whileServersAreRestarted() throws Exception {
+    createSessions();
+
+    AtomicBoolean running = new AtomicBoolean(true);
+    AtomicReference<String> phase = new AtomicReference<>("STARTUP");
+
+    Future<Integer> future1 = executor.submit(() -> sessionUpdater(0, running, phase));
+    Future<Integer> future2 = executor.submit(() -> sessionUpdater(1, running, phase));
+
+    GeodeAwaitility.await().during(1, TimeUnit.SECONDS).until(() -> true);
+
+    phase.set("CRASH 1 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 1 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 2 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 2 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("CRASH 3 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 3 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 4 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 4 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("FINISHING");
+    GeodeAwaitility.await().during(10, TimeUnit.SECONDS).until(() -> true);
+
+    running.set(false);
+    future1.get();
+    future2.get();
+  }
+
+  private Integer sessionUpdater(int index, AtomicBoolean running, AtomicReference<String> phase) {
+    int count = 0;
+    while (running.get()) {
+      int modCount = count % NUM_SESSIONS;
+      String sessionId = sessionIds.get(modCount);
+      Session session = findSession(sessionId);
+      assertThat(session).as("Session " + sessionId + " not found during phase " + phase.get())
+          .isNotNull();
+
+      session.setAttribute(String.format("attr-%d-%d", index, modCount), "value-" + count);
+      saveSession(session);
+      count++;
+    }
+
+    return count;
+  }
+
+  private Session findSession(String sessionId) {
+    Throwable latestException = null;
+
+    for (int i = 0; i < 10; i++) {
+      try {
+        return sessionRepository.findById(sessionId);
+      } catch (RedisSystemException rex) {
+        latestException = rex;
+        try {
+          Thread.sleep(2000);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+
+    throw new RuntimeException("Failed to find session after 10 attempts", latestException);
+  }
+
+  private void saveSession(Session session) {
+    Throwable latestException = null;
+
+    for (int i = 0; i < 10; i++) {
+      try {
+        sessionRepository.save(session);
+        return;
+      } catch (RedisSystemException rex) {
+        latestException = rex;
+        try {
+          Thread.sleep(2000);
```

Comment:
why is this needed and can it be less than 2 seconds? If not then you might need to make the GeodeAwaitility.during() (line 157) at least 20 seconds, because it's possible that we could spend that long trying to save the session and end up not going through all 10 attempts.

Reply:
Your comment prompted me to see if we can get away with just a single retry and it looks like the test is quite happy doing that.